examples/rust/tcp-push-pop.rs

```diff
@@ -100,218 +100,209 @@ fn close(libos: &mut LibOS, sockqd: QDesc) {
-        // Pop data.
-        let qt: QToken = match libos.pop(sockqd, None) {
+        // Accept incoming connections.
+        let qt: QToken = match self.libos.accept(self.sockqd) {
             Ok(qt) => qt,
-            Err(e) => {
-                // If error, free socket.
-                close(&mut libos, sockqd);
-                anyhow::bail!("pop failed: {:?}", e.cause)
-            },
+            Err(e) => anyhow::bail!("accept failed: {:?}", e.cause),
         };
-        let sga: demi_sgarray_t = match libos.wait(qt, None) {
-            Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_POP => unsafe { qr.qr_value.sga },
-            Err(e) => {
-                // If error, free socket.
-                close(&mut libos, sockqd);
-                anyhow::bail!("operation failed: {:?}", e.cause)
-            },
-            Ok(_) => {
-                // If error, free socket.
-                close(&mut libos, sockqd);
-                anyhow::bail!("unexpected result")
-            },
+
+        self.accepted_qd = match self.libos.wait(qt, None) {
+            Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_ACCEPT => unsafe { Some(qr.qr_value.ares.qd.into()) },
+            Err(e) => anyhow::bail!("operation failed: {:?}", e.cause),
+            Ok(_) => anyhow::bail!("unexpected result"),
         };
 
-        // Sanity check received data.
-        let ptr: *mut u8 = sga.sga_segs[0].sgaseg_buf as *mut u8;
-        let len: usize = sga.sga_segs[0].sgaseg_len as usize;
-        let slice: &mut [u8] = unsafe { slice::from_raw_parts_mut(ptr, len) };
-        for x in slice {
-            // If error, free socket.
-            if *x != FILL_CHAR {
-                close(&mut libos, sockqd);
-                anyhow::bail!("fill check failed: expected={:?} received={:?}", FILL_CHAR, *x);
+        // Perform multiple ping-pong rounds.
+        let mut i: usize = 0;
+        while i < nbytes {
+            // Pop data.
+            let qt: QToken = match self
+                .libos
+                .pop(self.accepted_qd.expect("should be a valid queue descriptor"), None)
+            {
+                Ok(qt) => qt,
+                Err(e) => anyhow::bail!("pop failed: {:?}", e.cause),
+            };
+
+            self.sga = match self.libos.wait(qt, None) {
+                Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_POP => unsafe { Some(qr.qr_value.sga) },
+                Err(e) => anyhow::bail!("operation failed: {:?}", e.cause),
+                Ok(_) => anyhow::bail!("unexpected result"),
+            };
+
+            // Sanity check received data.
+            let ptr: *mut u8 = self.sga.expect("should be a valid sgarray").sga_segs[0].sgaseg_buf as *mut u8;
+            let len: usize = self.sga.expect("should be a valid sgarray").sga_segs[0].sgaseg_len as usize;
+            let slice: &mut [u8] = unsafe { slice::from_raw_parts_mut(ptr, len) };
+
+            for x in slice {
+                if *x != fill_char {
```

Comment:
Can you change this to a demikernel::ensure_eq()?

Reply:
Done.